# ![](https://ga-dash.s3.amazonaws.com/production/assets/logo-9f88ae6c9c3871690e33280fcf557f33.png) Lesson 3.03 | Feature Engineering

## Review of Linear Regression

- Linear regression is a way for us to relate some dependent variable $Y$ to independent variables $X_1$,$\ldots$,$X_p$.
- We might write this out in one of the following two forms:
$$
\begin{eqnarray}
Y &=& \beta_0 + \beta_1X_1 + \beta_2X_2 + \cdots + \beta_pX_p +\varepsilon\\
\mathbf{Y} &=& \mathbf{X \beta + \varepsilon}
\end{eqnarray}
$$

<details><summary>
There are four assumptions to the simple linear regression model and five assumptions to the multiple linear regression model.
</summary>
1. **Linearity:** $Y$ is linearly related to $X_i$ for all $i$.
2. **Independence:** Each residual $\varepsilon_i$ is independent of $\varepsilon_j$ for all $i\neq j$.
3. **Normality:** The errors (residuals) follow a Normal distribution with mean 0.
4. **Equality of Variance:** The errors (residuals) should have a roughly consistent pattern, regardless of the value of $X_i$. (There should be no discernable relationship between $X_i$ and the residuals.)
5. **Independence Part II:** $X_i$ is independent of $X_j$ for all $i\neq j$.
</details>
- We can measure the performance of our model by using mean squared error (MSE).

## Feature Engineering

- If I use degrees Fahrenheit to predict how much a substance will expand or inches of rain to predict traffic accidents, people outside the United States may have a tougher time understanding my work.
- If I use straight line distance (as the crow flies) between two locations, my estimated time of arrival in a taxi or a Lyft is going to be pretty bad.
- If I put text into my model without some sort of preprocessing, my computer isn't going to understand how to handle it.

Suffice it to say: If your features (variables) aren't good, your predictions and inferences won't be good!

#### What is feature engineering?

"Coming up with features is difficult, time-consuming, requires expert knowledge. 'Applied machine learning' is basically feature engineering." - Andrew Ng

**Feature engineering** is the term broadly applied to the creation and manipulation of features (variables) used in machine learning algorithms.

Unless we're working with the same data over and over again, this isn't something we can automate. It will require creativity and a good, thorough understanding of our data.

#### The Process of Data Science
1. Data Gathering
2. Data Cleaning/Munging
3. EDA
4. Modeling
5. Reporting
    - Feature engineering will straddle all five of these steps, but mostly focus on steps 2 and 3.

#### [The Process of Feature Engineering](https://www.youtube.com/watch?v=drUToKxEAUA)
1. Brainstorming or testing features.
2. Deciding what features to create.
3. Creating features.
4. Checking how the features work with your model.
5. Improving features (if needed).
6. Return to step 1.
7. "Do data science!"

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("forestfires.csv")

In [3]:
df.head()

X  Y month  day  FFMC   DMC     DC  ISI  temp  RH  wind  rain  area
0  7  5   mar  fri  86.2  26.2   94.3  5.1   8.2  51   6.7   0.0   0.0
1  7  4   oct  tue  90.6  35.4  669.1  6.7  18.0  33   0.9   0.0   0.0
2  7  4   oct  sat  90.6  43.7  686.9  6.7  14.6  33   1.3   0.0   0.0
3  8  6   mar  fri  91.7  33.3   77.5  9.0   8.3  97   4.0   0.2   0.0
4  8  6   mar  sun  89.3  51.3  102.2  9.6  11.4  99   1.8   0.0   0.0

[Documentation](https://archive.ics.uci.edu/ml/machine-learning-databases/forest-fires/forestfires.names)

1. Number of Instances: 517 

2. Number of Attributes: 12 + output attribute
  
   Note: several of the attributes may be correlated, thus it makes sense to apply some sort of feature selection.

3. Attribute information:

   For more information, read [Cortez and Morais, 2007].

   - X - x-axis spatial coordinate within the Montesinho park map: 1 to 9
   - Y - y-axis spatial coordinate within the Montesinho park map: 2 to 9
   - month - month of the year: "jan" to "dec" 
   - day - day of the week: "mon" to "sun"
   - FFMC - FFMC index from the FWI system: 18.7 to 96.20
   - DMC - DMC index from the FWI system: 1.1 to 291.3 
   - DC - DC index from the FWI system: 7.9 to 860.6 
   - ISI - ISI index from the FWI system: 0.0 to 56.10
   - temp - temperature in Celsius degrees: 2.2 to 33.30
   - RH - relative humidity in %: 15.0 to 100
   - wind - wind speed in km/h: 0.40 to 9.40 
   - rain - outside rain in mm/m2 : 0.0 to 6.4 
   - area - the burned area of the forest (in ha): 0.00 to 1090.84 (this output variable is very skewed towards 0.0, thus it may make sense to model with the logarithm transform). 

In [4]:
import statsmodels.api as sm

/Users/mattbrems/anaconda3/lib/python3.5/site-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


What do each of the next six cells do? Chat in your local markets and we'll summarize together.

In [5]:
dep = df['area'] # Cell 1

In [6]:
indep = df.drop(['area','month','day'], axis = 'columns') # Cell 2

In [7]:
indep.head()

X  Y  FFMC   DMC     DC  ISI  temp  RH  wind  rain
0  7  5  86.2  26.2   94.3  5.1   8.2  51   6.7   0.0
1  7  4  90.6  35.4  669.1  6.7  18.0  33   0.9   0.0
2  7  4  90.6  43.7  686.9  6.7  14.6  33   1.3   0.0
3  8  6  91.7  33.3   77.5  9.0   8.3  97   4.0   0.2
4  8  6  89.3  51.3  102.2  9.6  11.4  99   1.8   0.0

In [8]:
indep = sm.add_constant(indep) # Cell 3

In [9]:
indep.head()

const  X  Y  FFMC   DMC     DC  ISI  temp  RH  wind  rain
0    1.0  7  5  86.2  26.2   94.3  5.1   8.2  51   6.7   0.0
1    1.0  7  4  90.6  35.4  669.1  6.7  18.0  33   0.9   0.0
2    1.0  7  4  90.6  43.7  686.9  6.7  14.6  33   1.3   0.0
3    1.0  8  6  91.7  33.3   77.5  9.0   8.3  97   4.0   0.2
4    1.0  8  6  89.3  51.3  102.2  9.6  11.4  99   1.8   0.0

In [10]:
model = sm.OLS(dep,indep) # Cell 4

In [11]:
results = model.fit() # Cell 5

In [13]:
results.summary() # Cell 6

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   area   R-squared:                       0.022
Model:                            OLS   Adj. R-squared:                  0.002
Method:                 Least Squares   F-statistic:                     1.119
Date:                Tue, 30 Jan 2018   Prob (F-statistic):              0.345
Time:                        10:45:34   Log-Likelihood:                -2874.8
No. Observations:                 517   AIC:                             5772.
Df Residuals:                     506   BIC:                             5818.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -6.3693     63.019     -0.101      0.920    -130.181     117.443
X              1.9079      1.448      1.317      0.188      -0.938       4.753
Y              0.5692      2.736      0.208      0.835      -4.807       5.945
FFMC          -0.0392      0.661     -0.059      0.953      -1.337       1.259
DMC            0.0773      0.067      1.151      0.250      -0.055       0.209
DC            -0.0033      0.016     -0.200      0.841      -0.036       0.029
ISI           -0.7137      0.772     -0.925      0.355      -2.229       0.802
temp           0.8002      0.787      1.017      0.310      -0.746       2.347
RH            -0.2306      0.237     -0.972      0.332      -0.697       0.236
wind           1.5574      1.670      0.933      0.352      -1.724       4.839
rain          -3.4040      9.681     -0.352      0.725     -22.424      15.616
==============================================================================
Omnibus:                      975.065   Durbin-Watson:                   1.647
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           781330.782
Skew:                          12.570   Prob(JB):                         0.00
Kurtosis:                     191.782   Cond. No.                     1.40e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.4e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [14]:
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                   area   R-squared:                       0.022
Model:                            OLS   Adj. R-squared:                  0.002
Method:                 Least Squares   F-statistic:                     1.119
Date:                Tue, 30 Jan 2018   Prob (F-statistic):              0.345
Time:                        10:45:40   Log-Likelihood:                -2874.8
No. Observations:                 517   AIC:                             5772.
Df Residuals:                     506   BIC:                             5818.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -6.3693     63.019     -0.101      0.9

Now, let's walk through some feature engineering!

#### Conversion

First, let's say that we want to convert `temp` from Celsius to Fahrenheit. Let's do that now by engineering a new feature `Fahr`, then build a new model and check the summary.

In [15]:
indep['Fahr'] = 1.8 * indep['temp'] + 32

In [16]:
indep.head()

const  X  Y  FFMC   DMC     DC  ISI  temp  RH  wind  rain   Fahr
0    1.0  7  5  86.2  26.2   94.3  5.1   8.2  51   6.7   0.0  46.76
1    1.0  7  4  90.6  35.4  669.1  6.7  18.0  33   0.9   0.0  64.40
2    1.0  7  4  90.6  43.7  686.9  6.7  14.6  33   1.3   0.0  58.28
3    1.0  8  6  91.7  33.3   77.5  9.0   8.3  97   4.0   0.2  46.94
4    1.0  8  6  89.3  51.3  102.2  9.6  11.4  99   1.8   0.0  52.52

In [17]:
model = sm.OLS(dep,indep)

In [18]:
results = model.fit()

In [20]:
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   area   R-squared:                       0.022
Model:                            OLS   Adj. R-squared:                  0.002
Method:                 Least Squares   F-statistic:                     1.119
Date:                Tue, 30 Jan 2018   Prob (F-statistic):              0.345
Time:                        10:52:22   Log-Likelihood:                -2874.8
No. Observations:                 517   AIC:                             5772.
Df Residuals:                     506   BIC:                             5818.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0711      0.276     -0.258      0.797      -0.613       0.471
X              1.9079      1.448      1.317      0.188      -0.938       4.753
Y              0.5692      2.736      0.208      0.835      -4.807       5.945
FFMC          -0.0392      0.661     -0.059      0.953      -1.337       1.259
DMC            0.0773      0.067      1.151      0.250      -0.055       0.209
DC            -0.0033      0.016     -0.200      0.841      -0.036       0.029
ISI           -0.7137      0.772     -0.925      0.355      -2.229       0.802
temp           1.1545      3.829      0.301      0.763      -6.369       8.678
RH            -0.2306      0.237     -0.972      0.332      -0.697       0.236
wind           1.5574      1.670      0.933      0.352      -1.724       4.839
rain          -3.4040      9.681     -0.352      0.725     -22.424      15.616
Fahr          -0.1968      1.961     -0.100      0.920      -4.049       3.656
==============================================================================
Omnibus:                      975.065   Durbin-Watson:                   1.647
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           781330.782
Skew:                          12.570   Prob(JB):                         0.00
Kurtosis:                     191.782   Cond. No.                     3.92e+18
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 1.31e-29. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

In [21]:
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                   area   R-squared:                       0.022
Model:                            OLS   Adj. R-squared:                  0.002
Method:                 Least Squares   F-statistic:                     1.119
Date:                Tue, 30 Jan 2018   Prob (F-statistic):              0.345
Time:                        10:52:29   Log-Likelihood:                -2874.8
No. Observations:                 517   AIC:                             5772.
Df Residuals:                     506   BIC:                             5818.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0711      0.276     -0.258      0.7

This is an example of a conversion. In this case, our goal isn't to add any new information to the model - we haven't given our model any "new" information - we just want to change the way our model accounts for temperature by using Fahrenheit!

**Check:** How would I interpret the coefficient for `temp` here? (Remember that our `Y` variable is `area`.)

#### Dummy Variables
Second, let's say that we want to convert `day` into a dummy variable. (A variable with values of `1`s and `0`s; enables our computer to understand categorical data.) There are a few ways of doing this. We'll go through one of the most common, `get_dummies()`.

In [22]:
df.head()

X  Y month  day  FFMC   DMC     DC  ISI  temp  RH  wind  rain  area
0  7  5   mar  fri  86.2  26.2   94.3  5.1   8.2  51   6.7   0.0   0.0
1  7  4   oct  tue  90.6  35.4  669.1  6.7  18.0  33   0.9   0.0   0.0
2  7  4   oct  sat  90.6  43.7  686.9  6.7  14.6  33   1.3   0.0   0.0
3  8  6   mar  fri  91.7  33.3   77.5  9.0   8.3  97   4.0   0.2   0.0
4  8  6   mar  sun  89.3  51.3  102.2  9.6  11.4  99   1.8   0.0   0.0

In [23]:
pd.get_dummies(df['day'])

fri  mon  sat  sun  thu  tue  wed
0      1    0    0    0    0    0    0
1      0    0    0    0    0    1    0
2      0    0    1    0    0    0    0
3      1    0    0    0    0    0    0
4      0    0    0    1    0    0    0
5      0    0    0    1    0    0    0
6      0    1    0    0    0    0    0
7      0    1    0    0    0    0    0
8      0    0    0    0    0    1    0
9      0    0    1    0    0    0    0
10     0    0    1    0    0    0    0
11     0    0    1    0    0    0    0
12     1    0    0    0    0    0    0
13     0    1    0    0    0    0    0
14     0    0    0    0    0    0    1
15     1    0    0    0    0    0    0
16     0    0    1    0    0    0    0
17     0    1    0    0    0    0    0
18     0    0    0    0    0    0    1
19     0    0    1    0    0    0    0
20     0    0    0    0    0    1    0
21     0    1    0    0    0    0    0
22     0    0    0    1    0    0    0
23     0    0    1    0    0    0    0
24     0    0    1    0    0    0    0
25     0    0    0    1    0    0    0
26     1    0    0    0    0    0    0
27     0    1    0    0    0    0    0
28     0    0    1    0    0    0    0
29     0    0    0    1    0    0    0
..   ...  ...  ...  ...  ...  ...  ...
487    0    0    0    0    0    1    0
488    0    0    0    0    0    1    0
489    0    0    0    0    0    0    1
490    0    0    0    0    0    0    1
491    0    0    0    0    1    0    0
492    1    0    0    0    0    0    0
493    1    0    0    0    0    0    0
494    0    0    1    0    0    0    0
495    0    1    0    0    0    0    0
496    0    1    0    0    0    0    0
497    0    0    0    0    0    1    0
498    0    0    0    0    0    1    0
499    0    0    0    0    0    1    0
500    0    0    0    0    0    1    0
501    0    0    0    0    0    1    0
502    0    0    0    0    0    1    0
503    0    0    0    0    0    0    1
504    0    0    0    0    0    0    1
505    0    0    0    0    1    0    0
506    1    0    0    0    0    0    0
507    1    0    0    0    0    0    0
508    1    0    0    0    0    0    0
509    1    0    0    0    0    0    0
510    1    0    0    0    0    0    0
511    0    0    0    1    0    0    0
512    0    0    0    1    0    0    0
513    0    0    0    1    0    0    0
514    0    0    0    1    0    0    0
515    0    0    1    0    0    0    0
516    0    0    0    0    0    1    0

[517 rows x 7 columns]

In [24]:
df.head()

X  Y month  day  FFMC   DMC     DC  ISI  temp  RH  wind  rain  area
0  7  5   mar  fri  86.2  26.2   94.3  5.1   8.2  51   6.7   0.0   0.0
1  7  4   oct  tue  90.6  35.4  669.1  6.7  18.0  33   0.9   0.0   0.0
2  7  4   oct  sat  90.6  43.7  686.9  6.7  14.6  33   1.3   0.0   0.0
3  8  6   mar  fri  91.7  33.3   77.5  9.0   8.3  97   4.0   0.2   0.0
4  8  6   mar  sun  89.3  51.3  102.2  9.6  11.4  99   1.8   0.0   0.0

In [25]:
pd.concat([df, pd.get_dummies(df['day'])], axis = 1)

X  Y month  day  FFMC    DMC     DC   ISI  temp  RH  wind  rain   area  \
0    7  5   mar  fri  86.2   26.2   94.3   5.1   8.2  51   6.7   0.0   0.00   
1    7  4   oct  tue  90.6   35.4  669.1   6.7  18.0  33   0.9   0.0   0.00   
2    7  4   oct  sat  90.6   43.7  686.9   6.7  14.6  33   1.3   0.0   0.00   
3    8  6   mar  fri  91.7   33.3   77.5   9.0   8.3  97   4.0   0.2   0.00   
4    8  6   mar  sun  89.3   51.3  102.2   9.6  11.4  99   1.8   0.0   0.00   
5    8  6   aug  sun  92.3   85.3  488.0  14.7  22.2  29   5.4   0.0   0.00   
6    8  6   aug  mon  92.3   88.9  495.6   8.5  24.1  27   3.1   0.0   0.00   
7    8  6   aug  mon  91.5  145.4  608.2  10.7   8.0  86   2.2   0.0   0.00   
8    8  6   sep  tue  91.0  129.5  692.6   7.0  13.1  63   5.4   0.0   0.00   
9    7  5   sep  sat  92.5   88.0  698.6   7.1  22.8  40   4.0   0.0   0.00   
10   7  5   sep  sat  92.5   88.0  698.6   7.1  17.8  51   7.2   0.0   0.00   
11   7  5   sep  sat  92.8   73.2  713.0  22.6  19.3  38   4.0   0.0   0.00   
12   6  5   aug  fri  63.5   70.8  665.3   0.8  17.0  72   6.7   0.0   0.00   
13   6  5   sep  mon  90.9  126.5  686.5   7.0  21.3  42   2.2   0.0   0.00   
14   6  5   sep  wed  92.9  133.3  699.6   9.2  26.4  21   4.5   0.0   0.00   
15   6  5   sep  fri  93.3  141.2  713.9  13.9  22.9  44   5.4   0.0   0.00   
16   5  5   mar  sat  91.7   35.8   80.8   7.8  15.1  27   5.4   0.0   0.00   
17   8  5   oct  mon  84.9   32.8  664.2   3.0  16.7  47   4.9   0.0   0.00   
18   6  4   mar  wed  89.2   27.9   70.8   6.3  15.9  35   4.0   0.0   0.00   
19   6  4   apr  sat  86.3   27.4   97.1   5.1   9.3  44   4.5   0.0   0.00   
20   6  4   sep  tue  91.0  129.5  692.6   7.0  18.3  40   2.7   0.0   0.00   
21   5  4   sep  mon  91.8   78.5  724.3   9.2  19.1  38   2.7   0.0   0.00   
22   7  4   jun  sun  94.3   96.3  200.0  56.1  21.0  44   4.5   0.0   0.00   
23   7  4   aug  sat  90.2  110.9  537.4   6.2  19.5  43   5.8   0.0   0.00   
24   7  4   aug  sat  93.5  139.4  594.2  20.3  23.7  32   5.8   0.0   0.00   
25   7  4   aug  sun  91.4  142.4  601.4  10.6  16.3  60   5.4   0.0   0.00   
26   7  4   sep  fri  92.4  117.9  668.0  12.2  19.0  34   5.8   0.0   0.00   
27   7  4   sep  mon  90.9  126.5  686.5   7.0  19.4  48   1.3   0.0   0.00   
28   6  3   sep  sat  93.4  145.4  721.4   8.1  30.2  24   2.7   0.0   0.00   
29   6  3   sep  sun  93.5  149.3  728.6   8.1  22.8  39   3.6   0.0   0.00   
..  .. ..   ...  ...   ...    ...    ...   ...   ...  ..   ...   ...    ...   
487  5  4   aug  tue  95.1  141.3  605.8  17.7  26.4  34   3.6   0.0  16.40   
488  4  4   aug  tue  95.1  141.3  605.8  17.7  19.4  71   7.6   0.0  46.70   
489  4  4   aug  wed  95.1  141.3  605.8  17.7  20.6  58   1.3   0.0   0.00   
490  4  4   aug  wed  95.1  141.3  605.8  17.7  28.7  33   4.0   0.0   0.00   
491  4  4   aug  thu  95.8  152.0  624.1  13.8  32.4  21   4.5   0.0   0.00   
492  1  3   aug  fri  95.9  158.0  633.6  11.3  32.4  27   2.2   0.0   0.00   
493  1  3   aug  fri  95.9  158.0  633.6  11.3  27.5  29   4.5   0.0  43.32   
494  6  6   aug  sat  96.0  164.0  643.0  14.0  30.8  30   4.9   0.0   8.59   
495  6  6   aug  mon  96.2  175.5  661.8  16.8  23.9  42   2.2   0.0   0.00   
496  4  5   aug  mon  96.2  175.5  661.8  16.8  32.6  26   3.1   0.0   2.77   
497  3  4   aug  tue  96.1  181.1  671.2  14.3  32.3  27   2.2   0.0  14.68   
498  6  5   aug  tue  96.1  181.1  671.2  14.3  33.3  26   2.7   0.0  40.54   
499  7  5   aug  tue  96.1  181.1  671.2  14.3  27.3  63   4.9   6.4  10.82   
500  8  6   aug  tue  96.1  181.1  671.2  14.3  21.6  65   4.9   0.8   0.00   
501  7  5   aug  tue  96.1  181.1  671.2  14.3  21.6  65   4.9   0.8   0.00   
502  4  4   aug  tue  96.1  181.1  671.2  14.3  20.7  69   4.9   0.4   0.00   
503  2  4   aug  wed  94.5  139.4  689.1  20.0  29.2  30   4.9   0.0   1.95   
504  4  3   aug  wed  94.5  139.4  689.1  20.0  28.9  29   4.9   0.0  49.59   
505  1  2   aug  thu  91.0  163.2  744.4  10.1  26.7  3

In [26]:
df = pd.concat([df, pd.get_dummies(df['day'])], axis = 1)

In [28]:
df.head()

X  Y month  day  FFMC   DMC     DC  ISI  temp  RH  wind  rain  area  fri  \
0  7  5   mar  fri  86.2  26.2   94.3  5.1   8.2  51   6.7   0.0   0.0    1   
1  7  4   oct  tue  90.6  35.4  669.1  6.7  18.0  33   0.9   0.0   0.0    0   
2  7  4   oct  sat  90.6  43.7  686.9  6.7  14.6  33   1.3   0.0   0.0    0   
3  8  6   mar  fri  91.7  33.3   77.5  9.0   8.3  97   4.0   0.2   0.0    1   
4  8  6   mar  sun  89.3  51.3  102.2  9.6  11.4  99   1.8   0.0   0.0    0   

   mon  sat  sun  thu  tue  wed  
0    0    0    0    0    0    0  
1    0    0    0    0    1    0  
2    0    1    0    0    0    0  
3    0    0    0    0    0    0  
4    0    0    1    0    0    0

In [30]:
dep = df['area']
indep = df.drop(['area','month','day'], axis = 'columns')
indep = sm.add_constant(indep)
model = sm.OLS(dep,indep)
results = model.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   area   R-squared:                       0.030
Model:                            OLS   Adj. R-squared:                 -0.001
Method:                 Least Squares   F-statistic:                    0.9576
Date:                Tue, 30 Jan 2018   Prob (F-statistic):              0.503
Time:                        11:13:14   Log-Likelihood:                -2872.6
No. Observations:                 517   AIC:                             5779.
Df Residuals:                     500   BIC:                             5852.
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -8.7901     55.322     -0.159      0.874    -117.482      99.902
X              1.8113      1.454      1.246      0.213      -1.045       4.668
Y              0.6610      2.750      0.240      0.810      -4.742       6.064
FFMC          -0.0184      0.665     -0.028      0.978      -1.325       1.288
DMC            0.0709      0.068      1.045      0.297      -0.062       0.204
DC            -0.0010      0.017     -0.059      0.953      -0.034       0.032
ISI           -0.6270      0.782     -0.802      0.423      -2.164       0.910
temp           0.7312      0.803      0.911      0.363      -0.846       2.308
RH            -0.2167      0.244     -0.890      0.374      -0.695       0.262
wind           1.8266      1.691      1.080      0.281      -1.496       5.150
rain          -3.2497      9.796     -0.332      0.740     -22.496      15.996
fri           -7.6974     10.424     -0.738      0.461     -28.178      12.783
mon           -2.9028     10.561     -0.275      0.784     -23.652      17.847
sat           10.9490     10.202      1.073      0.284      -9.095      30.993
sun           -3.2577     10.087     -0.323      0.747     -23.075      16.560
thu           -0.2100     10.944     -0.019      0.985     -21.712      21.292
tue           -1.0429     10.695     -0.098      0.922     -22.056      19.970
wed           -4.6282     11.427     -0.405      0.686     -27.079      17.822
==============================================================================
Omnibus:                      968.536   Durbin-Watson:                   1.638
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           750384.304
Skew:                          12.400   Prob(JB):                         0.00
Kurtosis:                     187.984   Cond. No.                     4.35e+18
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 1.05e-29. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

In [31]:
dep = df['area']
indep = df.drop(['area','month','day','fri'], axis = 'columns')
indep = sm.add_constant(indep)
model = sm.OLS(dep,indep)
results = model.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   area   R-squared:                       0.030
Model:                            OLS   Adj. R-squared:                 -0.001
Method:                 Least Squares   F-statistic:                    0.9576
Date:                Tue, 30 Jan 2018   Prob (F-statistic):              0.503
Time:                        11:13:39   Log-Likelihood:                -2872.6
No. Observations:                 517   AIC:                             5779.
Df Residuals:                     500   BIC:                             5852.
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -16.4875     63.745     -0.259      0.796    -141.728     108.753
X              1.8113      1.454      1.246      0.213      -1.045       4.668
Y              0.6610      2.750      0.240      0.810      -4.742       6.064
FFMC          -0.0184      0.665     -0.028      0.978      -1.325       1.288
DMC            0.0709      0.068      1.045      0.297      -0.062       0.204
DC            -0.0010      0.017     -0.059      0.953      -0.034       0.032
ISI           -0.6270      0.782     -0.802      0.423      -2.164       0.910
temp           0.7312      0.803      0.911      0.363      -0.846       2.308
RH            -0.2167      0.244     -0.890      0.374      -0.695       0.262
wind           1.8266      1.691      1.080      0.281      -1.496       5.150
rain          -3.2497      9.796     -0.332      0.740     -22.496      15.996
mon            4.7947     10.326      0.464      0.643     -15.494      25.083
sat           18.6464      9.941      1.876      0.061      -0.884      38.177
sun            4.4397      9.634      0.461      0.645     -14.488      23.368
thu            7.4874     10.931      0.685      0.494     -13.989      28.964
tue            6.6545     10.639      0.626      0.532     -14.247      27.556
wed            3.0692     11.234      0.273      0.785     -19.003      25.141
==============================================================================
Omnibus:                      968.536   Durbin-Watson:                   1.638
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           750384.304
Skew:                          12.400   Prob(JB):                         0.00
Kurtosis:                     187.984   Cond. No.                     1.41e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.41e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**Check:** How would I interpret the coefficient for `Mon`?

#### Advanced Feature Engineering
There's not an exhaustive list of ways we can engineer features. However, let's chat about a use-case in the context of this data.

We have horizontal and vertical spatial coordinates, listed as `X` and `Y` in our data. Perhaps we believe that area (our dependent variable) is linearly related to `X` or `Y`, but that's not very likely.

What if we think that `X` and `Y` have some combined effect on area? Here, we can create an **[interaction term](http://statisticsbyjim.com/regression/interaction-effects/)**.
- An interaction term is a way for us to account for the case when two independent variables have some joint effect or interact in some way.

In [33]:
dep = df['area']
indep = df[['X','Y']] ## EDITED
indep = sm.add_constant(indep)
indep['XY'] = indep['X'] * indep['Y'] ## NEW LINE
model = sm.OLS(dep,indep)
results = model.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   area   R-squared:                       0.006
Model:                            OLS   Adj. R-squared:                  0.000
Method:                 Least Squares   F-statistic:                     1.028
Date:                Tue, 30 Jan 2018   Prob (F-statistic):              0.380
Time:                        11:27:19   Log-Likelihood:                -2878.9
No. Observations:                 517   AIC:                             5766.
Df Residuals:                     513   BIC:                             5783.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         18.7297     19.784      0.947      0.344     -20.137      57.596
X             -1.8621      3.794     -0.491      0.624      -9.316       5.592
Y             -3.3095      5.033     -0.658      0.511     -13.198       6.579
XY             0.7887      0.819      0.963      0.336      -0.820       2.397
==============================================================================
Omnibus:                      981.720   Durbin-Watson:                   1.658
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           806473.069
Skew:                          12.751   Prob(JB):                         0.00
Kurtosis:                     194.801   Cond. No.                         198.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

If $X_1$ increases by 1, what is the effect on $Y$?
$$
\begin{eqnarray*}
Y_1 &=& \beta_0 + \beta_1X_1 + \beta_2X_2 + \beta_3X_1X_2 \\
\Rightarrow Y_2 &=& \beta_0 + \beta_1(X_1 + 1) + \beta_2X_2 + \beta_3(X_1 + 1)X_2 \\
\Rightarrow Y_\Delta &=& Y_2 - Y_1 \\
&=& \left(\beta_0 + \beta_1(X_1 + 1) + \beta_2X_2 + \beta_3(X_1 + 1)X_2\right) - \left(\beta_0 + \beta_1X_1 + \beta_2X_2 + \beta_3X_1X_2\right) \\
&=& (\beta_0 - \beta_0) + (\beta_1X_1 + \beta_1 - \beta_1X_1) + (\beta_2X_2 - \beta_2X_2) + (\beta_3X_1X_2 + \beta_3X_2 - \beta_3X_1X_2) \\
&=& \beta_1 + \beta_3X_2
\end{eqnarray*}
$$
If $X_1$ increases by 1, $Y$ is expected to increase by $\beta_1 + \beta_3X_2$.

Interpreting interaction terms is much less straightforward than interpreting "marginal" terms.

---

The last feature we'll engineer has to do with distance.

We have geographic data across a grid. We can talk about how horizontal distance matters or how vertical distance matters. We can also talk about how horizontal and vertical interact, but this is still clunky and awkward for us to handle.

Instead, what if we could actually define **distance** from a certain point?

$$
dist(A,B) = \sqrt{(A_X - B_X) ^ 2 + (A_Y - B_Y) ^ 2}
$$

Let's say that there's a specific landmark of interest at the point $X = 6, Y = 5$, and we want to engineer a new feature that represents distance from this landmark.

In [34]:
indep.head()

const  X  Y  XY
0    1.0  7  5  35
1    1.0  7  4  28
2    1.0  7  4  28
3    1.0  8  6  48
4    1.0  8  6  48

In [35]:
indep['Distance'] = ((indep['X'] - 6) ** 2 + (indep['Y'] - 5) ** 2) ** 0.5

In [36]:
indep.head(20)

const  X  Y  XY  Distance
0     1.0  7  5  35  1.000000
1     1.0  7  4  28  1.414214
2     1.0  7  4  28  1.414214
3     1.0  8  6  48  2.236068
4     1.0  8  6  48  2.236068
5     1.0  8  6  48  2.236068
6     1.0  8  6  48  2.236068
7     1.0  8  6  48  2.236068
8     1.0  8  6  48  2.236068
9     1.0  7  5  35  1.000000
10    1.0  7  5  35  1.000000
11    1.0  7  5  35  1.000000
12    1.0  6  5  30  0.000000
13    1.0  6  5  30  0.000000
14    1.0  6  5  30  0.000000
15    1.0  6  5  30  0.000000
16    1.0  5  5  25  1.000000
17    1.0  8  5  40  2.000000
18    1.0  6  4  24  1.000000
19    1.0  6  4  24  1.000000

## Conclusion

#### When do we decide when to stop?
- There's never infinite time.
- We have to consider **both** a) how useful it is and b) how much time you're willing to invest.
    - You will **always** be able to engineer more features.
    - [Diminishing marginal returns](https://www.investopedia.com/terms/l/lawofdiminishingmarginalreturn.asp) will absolutely play a role here.